src/async/smtp/MCSMTPSendWithDataOperation.cpp

```diff
@@ -84,3 +84,9 @@ void SMTPSendWithDataOperation::main()
     }
     setError(error);
 }
+
+void SMTPSendWithDataOperation::cancel()
+{
+    SMTPOperation::cancel();
+    session()->session()->cancelMessageSending();
```

Comment:
It's not clear why this additional check is required. In case if we call SMTPSendWithDataOperation::cancel before operation execution in serial queue, OperationQueue will filter this operation in OperationQueue::runOperations(), in other cases we have check in SMTPSession::internalSendMessage. Probably we can add additional check before loginIfNeeded(pError) to prevent unnecessary request.